p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,242 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"os"
+	"sync"
+
+	ic "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/pion/webrtc/v3"
+)
+
+var _ tpt.CapableConn = &connection{}
+
+const (
+	maxStreamBufferSize int = 10
+)
+
+type connection struct {
+	pc        *webrtc.PeerConnection
+	transport *WebRTCTransport
+	scope     network.ConnManagementScope
+
+	localPeer      peer.ID
+	privKey        ic.PrivKey
+	localMultiaddr ma.Multiaddr
+
+	remotePeer      peer.ID
+	remoteKey       ic.PubKey
+	remoteMultiaddr ma.Multiaddr
+
+	m       sync.Mutex
+	streams map[uint16]*dataChannel
+
+	streamChan chan network.MuxedStream
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newConnection(
+	pc *webrtc.PeerConnection,
+	transport *WebRTCTransport,
+	scope network.ConnManagementScope,
+
+	localPeer peer.ID,
+	privKey ic.PrivKey,
+	localMultiaddr ma.Multiaddr,
+
+	remotePeer peer.ID,
+	remoteKey ic.PubKey,
+	remoteMultiaddr ma.Multiaddr,
+) *connection {
+	streamChan := make(chan network.MuxedStream, maxStreamBufferSize)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	conn := &connection{
+		pc:        pc,
+		transport: transport,
+		scope:     scope,
+
+		localPeer:      localPeer,
+		privKey:        privKey,
+		localMultiaddr: localMultiaddr,
+
+		remotePeer:      remotePeer,
+		remoteKey:       remoteKey,
+		remoteMultiaddr: remoteMultiaddr,
+		ctx:             ctx,
+		cancel:          cancel,
+		streams:         make(map[uint16]*dataChannel),
+
+		streamChan: streamChan,
+	}
+
+	pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		id := *dc.ID()
+		var stream *dataChannel
+		dc.OnOpen(func() {
+			// datachannel cannot be detached before opening
+			rwc, err := dc.Detach()
+			if err != nil {
+				log.Errorf("[%s] could not detch channel: %s", localPeer, dc.Label())
+				return
+			}
+			stream = newDataChannel(conn, dc, rwc, pc, nil, nil)
+			conn.addStream(id, stream)
+			streamChan <- stream
```

Comment:
No, this does not block the pion event loop. The remote can create upto the maximum number of datachannels and pion will run the `onOpen` handler for all in separate goroutines. A solution could be to force close the incoming sctp stream if the datachannel is full.

Reply:
We need to avoid having 64k Go routines (in the worst case) sitting there and waiting to be able to send on this channel. Instead, we should reset streams if this channel fills up. That's a primitive way of creating backpressure, but that's all we can do given the constraints of the protocol and of Pion.